pkg/user/parallel_tx_submission.go

```diff
@@ -0,0 +1,497 @@
+	return balance.Amount.Int64() < DefaultWorkerBalance, nil
+}
+
+// hasFeeGrant checks if a fee grant exists between granter and grantee
+func (client *TxClient) hasFeeGrant(ctx context.Context, granter, grantee sdktypes.AccAddress) (bool, error) {
+	feegrantQuery := feegrant.NewQueryClient(client.conns[0])
+	_, err := feegrantQuery.Allowance(ctx, &feegrant.QueryAllowanceRequest{
+		Granter: granter.String(),
+		Grantee: grantee.String(),
+	})
+	if err != nil {
+		// If error contains "not found" or similar, grant doesn't exist
+		return false, nil
+	}
+	return true, nil
+}
+
+// fundAndGrantWorkerAccounts sends funds to worker accounts and sets up fee grants
+func (client *TxClient) fundAndGrantWorkerAccounts(ctx context.Context, workers []*txWorker) error {
+	if len(workers) == 0 {
+		return nil
+	}
+
+	msgs := make([]sdktypes.Msg, 0, len(workers)*2) // Each worker needs up to 2 msgs: send + feegrant
+	totalGasLimit := uint64(0)
+
+	masterAddress := client.defaultAddress
+
+	// Create send messages for funding accounts that need funding
+	for _, worker := range workers {
+		// Get worker address
+		record, err := client.signer.keys.Key(worker.accountName)
+		if err != nil {
+			return fmt.Errorf("failed to get worker account %s from keyring: %w", worker.accountName, err)
+		}
+
+		workerAddress, err := record.GetAddress()
+		if err != nil {
+			return fmt.Errorf("failed to get address for worker account %s: %w", worker.accountName, err)
+		}
+
+		// Check if account already has sufficient balance
+		needsFunding, err := client.accountNeedsFunding(ctx, workerAddress)
+		if err != nil {
+			// If we can't check balance, assume it needs funding to avoid blocking
+			needsFunding = true
+		}
+
+		if needsFunding {
+			// Create send message to fund the account
+			sendMsg := bank.NewMsgSend(
+				masterAddress,
+				workerAddress,
+				sdktypes.NewCoins(sdktypes.NewInt64Coin(appconsts.BondDenom, DefaultWorkerBalance)),
+			)
+			msgs = append(msgs, sendMsg)
+			totalGasLimit += SendGasLimit
+		}
+	}
+
+	// Add fee grant messages
+	feeGrantMsgs, feeGrantGasLimit, err := client.createFeeGrantMessages(ctx, workers)
+	if err != nil {
+		return err
+	}
+	msgs = append(msgs, feeGrantMsgs...)
+	totalGasLimit += feeGrantGasLimit
+
+	// Submit the initialization transaction only if there are messages to send
+	if len(msgs) > 0 {
+		opts := []TxOption{SetGasLimit(totalGasLimit)}
+		_, err = client.SubmitTx(ctx, msgs, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to submit initialization transaction: %w", err)
+		}
+	}
+
+	// Add the worker accounts to the signer
+	for _, worker := range workers {
+		record, err := client.signer.keys.Key(worker.accountName)
+		if err != nil {
+			return fmt.Errorf("failed to get worker account %s from keyring: %w", worker.accountName, err)
+		}
+
+		workerAddress, err := record.GetAddress()
+		if err != nil {
+			return fmt.Errorf("failed to get address for worker account %s: %w", worker.accountName, err)
```

Comment:
Similar logic seems to be written twice in the function maybe we can reduce duplication